proposals/lip-0039.md

```diff
@@ -302,273 +328,365 @@ proof={siblingHashes:[cc956a85, 3c516152, e041e1c0, 6400721e, 3c32b131],
-            remove p from queries
+```python
+def is_inclusion_proof(query_key: bytes, query: Query) -> bool:
+    return query_key == query.key and query.value != EMPTY_VALUE
+```
 
-    return queries
+```python
+def verify_inclusion_proof(query_keys: list[bytes], proof: Proof, root: bytes) -> bool:
+    # Check that there is a query for each key.
+    if len(query_keys) != len(proof.queries):
+        return False
+
+    for key, query in zip(query_keys, proof.queries):
+        # This is the only difference between an inclusion and a non-inclusion proof.
+        if is_inclusion_proof(key, query) == False:
+            return False
+
+    return verify(query_keys, proof, root)
 ```
 
-```java
-// This function checks whether two queries correspond to nodes that are
-// children of the same branch node, with q1 and q2 the left and right
-// child respectively
-function areSiblingQueries(q1, q2):
-    if length of q1.binaryBitmap != length of q2.binaryBitmap:
-        return false
-    let h=length of q1.binaryBitmap
-    let binaryKey1=binaryExpansion(q1.key)
-    let binaryKey2=binaryExpansion(q2.key)
-    let keyPrefix1 be the first (h-1) digits of binaryKey1
-    let keyPrefix2 be the first (h-1) digits of binaryKey2
-    if keyPrefix1 != keyPrefix2:
-        return false
-    let d1 be the digit at position h of binaryKey1
-    let d2 be the digit at position h of binaryKey2
-    return d1==0 and d2==1
+```python
+def verify_non_inclusion_proof(query_keys: list[bytes], proof: Proof, root: bytes) -> bool:
+    # Check that there is a query for each key.
+    if len(query_keys) != len(proof.queries):
+        return False
+
+    for key, query in zip(query_keys, proof.queries):
+        # This is the only difference between an inclusion and a non-inclusion proof.
+        if is_inclusion_proof(key, query) == True:
+            return False
+            
+    return verify(query_keys, proof, root)
 ```
 
-```java
-function calculateRoot(siblingHashes, queries):
-    // we add the hash property to each query, used to recompute the tree root
-    for each query q in queries:
-        set q.binaryBitmap to binaryExpansion(q.bitmap) without leading zeros
-        // if value is empty, it is a defaultEmptyNode
-        if q.value is empty bytes:
-            set q.hash to EMPTY_HASH
+```python
+def verify(query_keys: list[bytes], proof: Proof, root: bytes) -> bool:
+    for key, query in zip(query_keys, proof.queries):
+        # The bitmap does not contain extra leading 0s.
+        if query.bitmap[0] == 0:
+            return False
+        if len(key) != KEY_LENGTH_BYTES:
+            return False
+
+        # q is an inclusion proof for k or a default empty node.
+        if key == query.key:
+            continue
+
+        # q is an inclusion proof for another leaf node
+        common_prefix = os.path.commonprefix([binary_expansion(query.key), binary_expansion(key)])
+        if len(query.binary_bitmap) > len(common_prefix):
```

Comment:
What would be a tree that shows that this condition is necessary? Maybe we could also discuss in person.